### PR TITLE
Compiler: generate doc comments

### PR DESCRIPTION
**Pull request by kevmoo**
Fixes https://github.com/google/protobuf.dart/issues/161

### CODE REVIEW

File: protoc_plugin/lib/src/shared.dart [hunk 1]
```diff
@@ -2,8 +2,61 @@
...
+
+    final bits = descriptor.sourceCodeInfo.location
+        .where((element) => element.path.toString() == path.toString())
+        .toList();
+
+    if (bits.length == 1) {
+      final match = bits.single;
+      if (match.hasLeadingComments()) {
+        return toDartComment(match.leadingComments);
+      }
```
Comment (osa1) [inline protoc_plugin/lib/src/shared.dart:29]:
```suggestion
      return toDartComment(match.leadingComments);
```

When the field `leadingComments` isn't present the field value will be the default value of the field which is empty string, and `toDartComment` handles empty strings. So no need to check for presence.

File: protoc_plugin/lib/src/shared.dart [hunk 2]
```diff
@@ -2,8 +2,61 @@
...
+    }
+
+    if (bits.length > 1) {
+      throw 'WTH? ${descriptor.sourceCodeInfo.toProto3Json()}';
```
Comment (osa1) [inline protoc_plugin/lib/src/shared.dart:33]: I wonder if we should just print a warning (to stderr, stdout is used to print the generated code) and continue. `SourceCodeInfo` stuff does not affect runtime behavior of the protos so maybe a malformed `SourceCodeInfo` shouldn't block the operation.

We should also rephrase this with something helpful. Maybe something like "proto definition at $span has multiple source code infos".

File: protoc_plugin/lib/src/shared.dart [hunk 3]
```diff
@@ -2,8 +2,61 @@
...
+      lines.removeLast();
+    }
+
+    return lines.map((e) => '/// $e'.trimRight()).join('\n');
+  }
+
+  return null;
+}
+
+final _leadingSpaces = RegExp('^ +');
```
Comment (osa1) [inline protoc_plugin/lib/src/shared.dart:62]: What about other whitespace characters? https://protobuf.dev/reference/protobuf/textformat-spec/#whitespace

I really dislike regexes for simple things like this, but maybe this is the easiest in Dart..

File: protoc_plugin/lib/src/shared.dart [hunk 4]
```diff
@@ -2,8 +2,61 @@
...
+    }
+
+    while (lines.last.isEmpty) {
+      lines.removeLast();
+    }
+
+    return lines.map((e) => '/// $e'.trimRight()).join('\n');
+  }
+
+  return null;
```
Comment (osa1) [inline protoc_plugin/lib/src/shared.dart:59]: If you move this check at the beginning with

```dart
if (value.isEmpty) return null;
```

Then the function will need less indentation and be easier to read. Same as in `commentBlock` where you check for the early return first.

File: protoc_plugin/pubspec.yaml [hunk 5]
```diff
@@ -1,5 +1,5 @@
 name: protoc_plugin
-version: 21.0.2
+version: 21.1.0-wip
```
Comment (osa1) [inline protoc_plugin/pubspec.yaml:2]:
```suggestion
version: 21.1.0-dev
```

The convention we use in protoc_plugin and protobuf is `-dev`. protobuf lib version currently is `3.0.1-dev`.

**Review by osa1 — APPROVED** [review]
Looks great, thanks @kevmoo. Added a few inline comments and suggestions.

File: protoc_plugin/CHANGELOG.md [hunk 6]
```diff
@@ -1,3 +1,9 @@
+## 21.1.0-wip
```
Comment (osa1) [inline protoc_plugin/CHANGELOG.md:1]:
```suggestion
## 21.1.0-dev
```

(Same as the other suggestion)

File: protoc_plugin/CHANGELOG.md [hunk 7]
```diff
@@ -1,3 +1,9 @@
...
+
+* Experimental support for generating comments. Use the `generate_comments` option. ([#161])
```
Comment (osa1) [inline protoc_plugin/CHANGELOG.md:3]: Why do you say it's experimental? If there isn't anything unclear (maybe regarding how the `SourceCodeInfo` etc. are used) or unsupported I think we should just advertise this as "support".

Comment (kevmoo) [reply protoc_plugin/CHANGELOG.md:3]: I'm just not very committal. We can call it "ready"

In fact, happy to discuss not putting it behind a flag.

Comment (osa1) [reply protoc_plugin/CHANGELOG.md:3]: I checked a few of the protoc compilers used internally. Java and Go generate comments. C++ doesn't. AFAIK this library mostly copies Java (the lite version) so I say we generate comments always and remove the flag.

File: protoc_plugin/lib/src/shared.dart (hunk 2 above)
Comment (kevmoo) [reply protoc_plugin/lib/src/shared.dart]: yup. fixed

File: protoc_plugin/lib/src/shared.dart (hunk 3 above)
Comment (osa1) [reply protoc_plugin/lib/src/shared.dart:59]: Yeah tabs should be common. I don't even know what vertical tab and form feed are and I suspect they're not used in practice. Carriage return should be handled by the line splitter.

Comment (kevmoo) [reply protoc_plugin/lib/src/shared.dart:59]: I think it's okay to handle these case-by-case. In all of the bits I looked through they ALWAYS seem to have one leading space – that's it!

**Comment by MattyBoy4444** [comment]
This doesn't grab comments at the end of the line.

EX: 

message Animal {
  string name = 1; //This is the name of the animal
}

**Comment by kevmoo** [comment]
File another bug on that

**Pull request merged**